### PR TITLE
Some bitwise and bool2Word lemmas

### PR DESCRIPTION
**Pull request by ehildenb**
This PR is the first of several that will come from some experiments with the Kontrol test-suite. This one adjusts the semantics and lemmas, in particular:

- In the process of exploring generated claims from Kontrol test-suite, several simplifications were revealed that seem generally useful. The simplifications are added along with tests:
  - A rule about simplifying `X xorInt Y <Int 0 -Int X`, if `X` and `Y` are both positive.
  - Rules about simplifying `X ==Int 0` when disequalities about `X` around zero are known.
  - Claims which were failing before and now are passing with these lemmas are added.

### CODE REVIEW

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k [hunk 1]
```diff
@@ -243,4 +243,11 @@ module INT-SIMPLIFICATION-COMMON
       requires 0 <Int Y
       [simplification, comm]
 
+  // ###########################################################################
+  // comparisons
+  // ###########################################################################
+
+    rule X ==Int 0 => true  requires 0 <=Int X andBool X <=Int 0 [simplification, comm]
+    rule X ==Int 0 => false requires 0 <Int X                    [simplification, comm]
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k:251]: The tests added are exactly the expressions that came up causing these lemmas to be added. I can try out the lemmas you suggest.

Whenever I run into something unsimplified that I want simplified, the _first_ thing I do is snip out that expression and add it as a test to KEVM. Then I begin working on that test directly.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k [hunk 2]
```diff
@@ -87,6 +87,12 @@ module BITWISE-SIMPLIFICATION [symbolic]
       lengthBytes(X) +Int 32 >Int notMaxUInt5 &Int ( lengthBytes(X) +Int 31 ) => true
       [simplification, smt-lemma]
 
+  // ###########################################################################
+  // bit-xor
+  // ###########################################################################
+
+    rule X xorInt Y <Int 0 -Int X => false requires 0 <=Int X andBool 0 <=Int Y [simplification]
```
[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k:94]: The tests added are coming directly from simplifications I wanted to happen which didn't, when producing this proof suite: https://github.com/runtimeverification/evm-semantics/pull/2244

I can try out the more general simplification.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k (hunk 1 above)
[the remark this answers is not in the document]
Comment (PetarMax) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/int-simplification.k:251]: Yes, I understand that. I'm not suggesting that you replace your lemmas with the ones I suggest. All that I am saying is that I had similar lemmas that were slowing down the performance. I suspect that your lemmas are causing a slowdown that resulted in the CI failure. Have you got a performance comparison of this branch vs master?

**Comment by PetarMax** [comment]
I think that we could benefit from introducing simplifications that only do a syntactic search of the pure constraints. For example:

`rule X ==Int 0 => true  requires 0 <=Int X andBool X <=Int 0 [simplification, comm, syntactic]`

would only fire if the requires clause conjucts were verbatim in the pure constraints. I think this is the case in the examples that motivated this particular simplification and I know it was the case when I introduced the more general ones.

File: kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k (hunk 2 above)
[the remark this answers is not in the document]
Comment (ehildenb) [reply kevm-pyk/src/kevm_pyk/kproj/evm-semantics/lemmas/bitwise-simplification.k]: Done.

**Comment by ehildenb** [comment]
It appears that all proofs take the same amount of time as master, _except_ the bihu functional specification here, which loops: https://github.com/runtimeverification/evm-semantics/blob/dca942634ed69288b58bab93831e4cb6dace75bd/tests/specs/bihu/functional-spec.k#L18.

I'm investigating.

**Comment by yale-vinson** [comment]
Any update on this one? I see from two weeks ago you are investigating, but curious to see if you made any headway on it?

**Comment by ehildenb** [comment]
Subsumed: https://github.com/runtimeverification/evm-semantics/pull/2270
Subsumed: https://github.com/runtimeverification/evm-semantics/pull/2253

**Pull request closed without merging**